server/chat-commands/avatars.tsx

```diff
@@ -0,0 +1,869 @@
+
+		if (!Avatars.addPersonal(userid, avatar)) {
+			throw new Chat.ErrorMessage(`User "${inputUsername}" can already use avatar "${avatar}".`);
+		}
+		this.globalModlog('PAVATAR', userid, avatar);
+		this.sendReplyBox(<div>
+			{Avatars.img(avatar)}<br />
+			Added to <username class="username">{inputUsername}</username>
+		</div>);
+	},
+	defaultavatarhelp: 'addavatarhelp',
+
+	allowedavatar: 'allowavatar',
+	groupavatar: 'allowavatar',
+	async allowavatar(target, room, user) {
+		this.checkCan('bypassall');
+		if (!target) return this.parse(`/help defaultavatar`);
+		const [inputUsername, inputAvatar] = this.splitOne(target);
+		if (!Users.isUsername(inputUsername)) {
+			throw new Chat.ErrorMessage(`"${inputUsername}" is not a valid username.`);
+		}
+		const userid = toID(inputUsername);
+		const avatar = await Avatars.validate(inputAvatar, {rejectOfficial: true});
+
+		if (!Avatars.addAllowed(userid, avatar)) {
+			throw new Chat.ErrorMessage(`User "${inputUsername}" can already use avatar "${avatar}".`);
+		}
+		this.globalModlog('GAVATAR', userid, avatar);
+		this.sendReplyBox(<div>
+			{Avatars.img(avatar)}<br />
+			Added to <username class="username">{inputUsername}</username>
+		</div>);
+	},
+	allowavatarhelp: 'addavatarhelp',
+
+	denyavatar: 'removeavatar',
+	disallowavatar: 'removeavatar',
+	removeavatar(target, room, user) {
+		this.checkCan('bypassall');
+		if (!target) return this.parse(`/help defaultavatar`);
+		const [inputUsername, inputAvatar] = this.splitOne(target);
+		if (!Users.isUsername(inputUsername)) {
+			throw new Chat.ErrorMessage(`"${inputUsername}" is not a valid username.`);
+		}
+		const userid = toID(inputUsername);
+		const avatar = Avatars.convert(inputAvatar);
+
+		const allowed = customAvatars[userid]?.allowed.filter(Boolean);
+		if (!allowed) {
+			throw new Chat.ErrorMessage(`${inputUsername} doesn't have any custom avatars.`);
+		}
+		if (avatar) {
+			if (!Avatars.removeAllowed(userid, avatar)) {
+				throw new Chat.ErrorMessage(`${inputUsername} doesn't have access to avatar "${avatar}"`);
+			}
+			this.globalModlog('REMOVE AVATAR', userid, avatar);
+			this.sendReplyBox(<div>
+				{Avatars.img(avatar)}<br />
+				Removed from <username class="username">{inputUsername}</username>
+			</div>);
+		} else {
+			// delete all
+			delete customAvatars[userid];
+			Avatars.save();
+			this.globalModlog('REMOVE AVATAR', userid, allowed.join(','));
+			this.sendReplyBox(<div>
+				{allowed.map(curAvatar => [Avatars.img(curAvatar!), ' '])}<br />
+				Removed from <username class="username">{inputUsername}</username>
+			</div>);
+		}
+	},
+	removeavatarhelp: 'addavatarhelp',
+
+	async avatarusers(target, room, user) {
+		target = '#' + toID(target);
+		if (!Avatars.userCanUse(user, target) && !user.can('alts')) {
+			throw new Chat.ErrorMessage(`You don't have access to avatar "${target}"`);
+		}
+
+		if (!this.runBroadcast()) return;
```

Comment:
Nitpick, but strictly speaking this isn't necessary since it uses ErrorMessage.
```suggestion
		this.runBroadcast();
```

Reply:
Oh huh this was a carry-over from the old command. We should probably port all the other uses forward instead of going one-by-one.